ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -1,24 +1,367 @@
+
+          // Add a mapping from the topic+datatype tuple to the channel, and channel ID to the
+          // topic+datatype tuple
+          _advertisedTopics.emplace(topicAndDatatype, std::move(channel));
+          _channelToTopicAndDatatype.emplace(channel.id, topicAndDatatype);
+        } catch (foxglove::DefinitionNotFoundError& err) {
+          RCLCPP_WARN(this->get_logger(), "Could not find definition for topic \"%s\" (%s)",
+                      topicAndDatatype.first.c_str(), topicAndDatatype.second.c_str());
+
+          // Add a mapping from the topic+datatype tuple to a dummy channel so we don't repeatedly
+          // try to load the message definition
+          auto channel = foxglove::Channel{0, newChannel};
+          _advertisedTopics.emplace(topicAndDatatype, std::move(channel));
+        }
+      }
+    }
+
+    if (newTopics.size() > 0) {
+      _server.broadcastChannels();
+    }
+
+    // Schedule the next update
+    _updateCount++;
+    auto nextUpdateMs =
+      std::chrono::milliseconds(std::min(size_t(1) << _updateCount, _maxUpdateMs));
+    _updateTimer = this->create_wall_timer(
+      nextUpdateMs, std::bind(&FoxgloveBridge::updateAdvertisedTopics, this));
   }
 
 private:
+  struct PairHash {
+    template <class T1, class T2>
+    std::size_t operator()(const std::pair<T1, T2>& pair) const {
+      return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+    }
+  };
+
+  foxglove::Server<foxglove::WebSocketNoTls> _server;
+  foxglove::MessageDefinitionCache _messageDefinitionCache;
+  std::unordered_map<TopicAndDatatype, foxglove::Channel, PairHash> _advertisedTopics;
+  std::unordered_map<foxglove::ChannelId, TopicAndDatatype> _channelToTopicAndDatatype;
+  std::unordered_map<foxglove::ChannelId, std::shared_ptr<rclcpp::GenericSubscription>>
+    _subscriptions;
+  std::mutex _subscriptionsMutex;
+  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr _parametersUpdateHandle;
+  rclcpp::TimerBase::SharedPtr _updateTimer;
+  size_t _maxUpdateMs = size_t(DEFAULT_MAX_UPDATE_MS);
+  size_t _updateCount = 0;
+
+  rcl_interfaces::msg::SetParametersResult parametersHandler(
+    const std::vector<rclcpp::Parameter>& parameters) {
+    for (const auto& parameter : parameters) {
+      if (parameter.get_name() == "max_update_ms") {
+        _maxUpdateMs = size_t(parameter.as_int());
+      }
+    }
+
+    rcl_interfaces::msg::SetParametersResult result;
+    result.successful = true;
+    result.reason = "success";
+    return result;
+  }
+
+  void subscribeHandler(foxglove::ChannelId channelId) {
+    std::lock_guard<std::mutex> lock(_subscriptionsMutex);
+    if (_subscriptions.find(channelId) != _subscriptions.end()) {
+      // Subscription already exists
+      RCLCPP_DEBUG(this->get_logger(), "Subscription already exists for channel %d", channelId);
+      return;
+    }
+
+    auto it = _channelToTopicAndDatatype.find(channelId);
```

Comment:
```suggestion
    const auto it = _channelToTopicAndDatatype.find(channelId);
```
also in some places below

Reply:
Why?